File: src/python/turtle/Turtle.py
from Vector import *
from OpenGL.GL import *
from OpenGL.GLUT import *
from OpenGL.arrays import vbo
import pdb
import math
import colorsys

from Conf import Conf
import random
import numpy as np
import time

class Turtle:
	"""This class is the turtle that will draw the fractals"""

	class State:
		"""This class represent a state of the turtle, for push and pop"""
		def __init__(self, pos, heading, color):
			self.pos = Vector().set(pos)
			self.heading = Vector().set(heading)
			self.color = Vector().set(color)

	def __init__(self, pos=None):
		if pos is None:
			self.pos = Vector(Conf.TURTLE.INIT_POS)
		else:
			self.pos = Vector(pos)
		self.heading = Vector(Conf.TURTLE.INIT_HEADING)
		self.color = Vector(colorsys.rgb_to_hls(Conf.TURTLE.INIT_COLOR[0], Conf.TURTLE.INIT_COLOR[1], Conf.TURTLE.INIT_COLOR[2]))

		self.stateStack = [] # allow to manage a stack of states
		self.stochasticFactor = 0 # used to randomize drawing

		# the vertex buffer is a liste of lists : [[x y z r g b time],...]
		self.vertexBuffer = []
		self.vertexBufferLength = 0
		self.vertexBufferChanged = False
		self.vertexPositions = None
		#indices = np.array([[0,1,2], [0, 3, 2]], dtype=np.int32)
		self.draw_type = GL_TRIANGLES

		self.point()

	""" allow to change the way the turtle is drawing the fractal """
	def setDrawType(self, type):
		self.draw_type = type

	""" Set the stochastic factor of this instance of the turtle """
	def setStochasticFactor(self, factor):
		self.stochasticFactor = factor

	""" Return a random number in a range around the given number. The range is defined by the stochastoc factor """
	def randomize(self, value):
		if self.stochasticFactor:
			return ((random.random() * 2 * self.stochasticFactor) - self.stochasticFactor) * value + value
		return value

	""" Reinit the turtle to its initial state """
	def reinit(self):
		self.pos = Vector(Conf.TURTLE.INIT_POS)
		self.heading = Vector(Conf.TURTLE.INIT_HEADING)
		self.color = Vector(colorsys.rgb_to_hls(Conf.TURTLE.INIT_COLOR[0], Conf.TURTLE.INIT_COLOR[1], Conf.TURTLE.INIT_COLOR[2]))
		self.vertexBuffer = []
		self.vertexBufferChanged = False
		self.vertexBufferLength = 0
		self.vertexPositions = vbo.VBO(np.array(self.vertexBuffer, dtype=np.float32))
		self.vertexPositions.bind()


	""" Draw a point at the current position of the turtle """
	def point(self, breakline = 0):
		c = colorsys.hls_to_rgb(self.color.x, self.color.y, self.color.z)
		self.vertexBuffer.append([float(self.pos.x), float(self.pos.y), float(self.pos.z),
							c[0], c[1], c[2], breakline])
		self.vertexBufferChanged = True
		self.vertexBufferLength += 1


	""" Begin a command list """
	def begin(self, reinit = False):
		# lorsque 'end()' est appele, si le vertexBuffer a change le vbo sera actualise
		self.vertexBufferChanged = False

		if reinit:
			self.reinit()
			glutSolidSphere(0.1, 10, 10)


	""" Push current state (pos, heading and color) on the state stack """
	def push(self, arg):
		if 'turtle' in Conf.DEBUG:
			print "Called: push"
		self.stateStack.append(Turtle.State(self.pos, self.heading, self.color))


	""" Rollback to the last pushed state of the state stack """
	def pop(self, arg):
		if 'turtle' in Conf.DEBUG:
			print "Called: pop"
		state = self.stateStack.pop()
		self.pos.set(state.pos)
		self.heading.set(state.heading)
		self.color.set(state.color)
		self.point(1)

	""" Move forward (draw a point) """
	def forward(self, step):
		if 'turtle' in Conf.DEBUG:
			print "Called: forward(", step, ");"
		#print "call: forward"
		self.pos += self.randomize(step) * self.heading
		self.point()

	""" Move backward (draw a point) """
	def backward(self, step):
		if 'turtle' in Conf.DEBUG:
			print "Called: backward(", step, ");"
		self.pos += ((self.randomize(step) * self.heading) * -1)
		self.point()

	""" Rotation around the x direction """
	def rotX(self, angle):
		if 'turtle' in Conf.DEBUG:
			print "Called: rotX(", angle, ");"
		angle = self.randomize(angle)
		self.heading.set((
			self.heading.x,
			self.heading.y * math.cos(angle) - self.heading.z * math.sin(angle),
			self.heading.y * math.sin(angle) + self.heading.z * math.cos(angle)))
	""" Reverse rotation around the x direction"""
	def irotX(self, angle):
		if 'turtle' in Conf.DEBUG:
			print "Called: rotX(", angle, ");"
		angle = self.randomize(angle)
		angle = -angle
		self.heading.set((
			self.heading.x,
			self.heading.y * math.cos(angle) - self.heading.z * math.sin(angle),
			self.heading.y * math.sin(angle) + self.heading.z * math.cos(angle)))

	""" Rotation around the y direction """
	def rotY(self, angle):
		if 'turtle' in Conf.DEBUG:
			print "Called: rotY(", angle, ");"
		angle = self.randomize(angle)
		self.heading.set((
			self.heading.x * math.cos(angle) + self.heading.z * math.sin(angle),
			self.heading.y,
			- self.heading.x * math.sin(angle) + self.heading.z * math.cos(angle)))
	""" Reverse rotation around the y direction"""
	def irotY(self, angle):
		if 'turtle' in Conf.DEBUG:
			print "Called: rotY(", angle, ");"
		angle = self.randomize(angle)
		angle = -angle
		self.heading.set((
			self.heading.x * math.cos(angle) + self.heading.z * math.sin(angle),
			self.heading.y,
			- self.heading.x * math.sin(angle) + self.heading.z * math.cos(angle)))

	""" Rotation around the z direction """
	def rotZ(self, angle):
		if 'turtle' in Conf.DEBUG:
			print "Called: rotZ(", angle, ");"
		angle = self.randomize(angle)
		self.heading.set((
			self.heading.x * math.cos(angle) - self.heading.y * math.sin(angle),
			self.heading.x * math.sin(angle) + self.heading.y * math.cos(angle),
			self.heading.z))
	""" Reverse rotation around the z direction"""
	def irotZ(self, angle):
		if 'turtle' in Conf.DEBUG:
			print "Called: rotZ(", angle, ");"
		angle = self.randomize(angle)
		angle = -angle
		self.heading.set((
			self.heading.x * math.cos(angle) - self.heading.y * math.sin(angle),
			self.heading.x * math.sin(angle) + self.heading.y * math.cos(angle),
			self.heading.z))

	def setColor(self, color):
		if 'turtle' in Conf.DEBUG:
			print "Called: setColor(", color, ");"
		#color = self.randomize(color)
		self.color.set(colorsys.rgb_to_hls(color[0], color[1], color[2]))

	def nextColor(self, step):
		if 'turtle' in Conf.DEBUG:
			print "Called: nextColor(", step, ");"
			print "Previous: ", self.color.toString()
		self.color.x += step
		if self.color.x > 1.0:
			self.color.x = self.color.x - 1.0
		if 'turtle' in Conf.DEBUG:
			print "Next: ", self.color.toString()
	""" Close a commend sequence """
	def end(self):
		glBegin(GL_LINES)
		glColor3f(0, 0, 1)
		glVertex3f(self.pos.x, self.pos.y, self.pos.z)
		glVertex3f(float(self.pos.x + self.heading.x), float(self.pos.y + self.heading.y), float(self.pos.z + self.heading.z));
		glEnd()

	""" Draw the turtle path """
	def draw(self):
		if self.vertexBufferChanged or self.vertexPositions is None:
			#print "Vertex buffer changed !"
			#print "Creating numpy array..."
			vertices = np.array(self.vertexBuffer, \
		 					dtype=np.float32)
			#print vertices
			#print "Creating VBO..."
			self.vertexPositions = vbo.VBO(vertices)
			#print "VBO: ", self.vertexPositions.data
			print "Total number of vertex: ", self.vertexBufferLength
			self.vertexBufferChanged = False



		#self.indexPositions.bind()
		self.vertexPositions.bind()
		glEnableClientState(GL_VERTEX_ARRAY)
		glEnableClientState(GL_COLOR_ARRAY)

		glVertexPointer(3, GL_FLOAT, 28, self.vertexPositions )
		glColorPointer(3, GL_FLOAT, 28, self.vertexPositions+12 )
		#glVertexAttribPointer(7, 1, GL_FLOAT, GL_TRUE, 28, self.vertexPositions+24)

		glDrawArrays(self.draw_type, 0, self.vertexBufferLength);
		#glDrawElements(GL_TRIANGLES, 6, GL_UNSIGNED_INT, None)

		#self.indexPositions.unbind()
		self.vertexPositions.unbind()
		glDisableClientState(GL_VERTEX_ARRAY)
		glDisableClientState(GL_COLOR_ARRAY)

File: src/js/turtle.js

/**
 * Implementation of a Logo Turtle
 * Used to execute a generated fractal's code
 */
class Turtle {
    /**
     * Initialize the turtle
     * @param {Graphx} graphx - graphx instance used to add meshes to the scene
     * @param {Object} [pos] - initial position of the turtle, as a {x, y, z} object (default to the origin)
     * @param {Boolean} [debug] - enable debugging mode
     */
    constructor(graphx, debug, pos) {
        this.debug = debug;
        this.graphx = graphx;
        this.stateStack = [];
        this.stochasticFactor = 0;

        this.drawType = 'line_strip';

        this.reinit(pos);
        this.point();
    }

    setDrawType(drawType) {
        this.drawType = drawType;
    }

    randomize(value) {
        if (!this.stochasticFactor) { return value; }
        return ((Math.random() * 2 * this.stochasticFactor) - this.stochasticFactor) * value + value;
    }

    reinit(pos) {
        if (!pos) {
            this.pos = new THREE.Vector3(0, 0, 0)
        }
        else {
            this.pos = new THREE.Vector3(pos.x, pos.y, pos.z)
        }

        this.heading = new THREE.Vector3(0, 1, 0);
        this.color = new THREE.Color(0.5, 0.25, 0.0);

        this.vertexBuffers = [];
        this.vertexBufferLen = 0;
        this.vertexBufferChanged = false;
        this.vertexPositions = null;

        if (this.meshes) {
            this.meshes.forEach(m => {
                this.graphx.scene.remove(m);
            });
        }
        this.meshes = [];
    }

    /**
     * Draw the turtle path
     */
    update() {
        if (this.vertexBufferChanged) {
            this.meshes.forEach(m => {
                this.graphx.scene.remove(m);
            });

            let total = 0;
            this.meshes = [this._miniSphere()].concat(this.vertexBuffers.map(vertexBuffer => {
                total += vertexBuffer.positions.length / 3;
                // TODO: honor this.drawStyle
                const geometry = new THREE.BufferGeometry();
                geometry.addAttribute('position', new THREE.Float32BufferAttribute(vertexBuffer.positions, 3));
                geometry.addAttribute('color', new THREE.Float32BufferAttribute(vertexBuffer.colors, 3));
                const material = new THREE.LineBasicMaterial({});
                return new THREE.Line(geometry, material);
            }));

            this.vertexBufferLen = total;

            this.vertexBufferChanged = false;

            this.meshes.forEach(m => {
                this.graphx.scene.add(m);
            });
        }
    }

    /**
     * Add a point at the current position of the turtle
     * @param {Boolean} breakline - enable to prevent a line to be drawn from the previous point
     */
    point(breakline=false) {
        if (breakline || this.vertexBuffers.length === 0) {
            this.vertexBuffers.push(new VertexBuffer());
        }
        const vertexBuffer = this.vertexBuffers[this.vertexBuffers.length - 1];

        vertexBuffer.addVertex(
            [this.pos.x, this.pos.y, this.pos.z],
            [this.color.r, this.color.g, this.color.b]
        );

        this.vertexBufferLen += 1;
        this.vertexBufferChanged = true;
    }

    /**
     * Begin a command list
     * @param {Boolean} reinit - enable to reinitialize the turtule
     */
    begin(reinit=false) {
        if (this.debug) {
            logging.log("[Turtle] begin(reinit=" + reinit + ")");
        }
        this.vertexBufferChanged = false;

        if (reinit) {
            this.reinit();
        }
    }

    /**
     * Push current state (pos, heading and color) onto the state stack
     */
    push() {
        if (this.debug) {
            logging.log("[Turtle] push()");
        }
        this.stateStack.push(new TurtleState(this.pos, this.heading, this.color));
    }

    /**
     * Rollback to the last pushed state of the state stack
     */
    pop() {
        if (this.debug) {
            logging.log("[Turtle] pop()");
        }
        const state = this.stateStack.pop();
        this.color.copy(state.color);
        this.pos.copy(state.pos);
        this.heading.copy(state.heading);
    }

    /**
     * Move forward (and draw a point)
     * @param {Float} step distance
     */
    forward(step) {
        if (this.debug) {
            logging.log('[Turtle] forward(' + step + ')');
        }
        this.pos.add(this.heading.clone().multiplyScalar(this.randomize(step)));
        this.point();
    }

    /**
     * Rotation around the X axis
     * @param {Float} angle in radians
     */
    rotX(angle) {
        if (this.debug) {
            logging.log('[Turtle] rotX(' + angle + ')');
        }
        angle = this.randomize(angle);
        this.heading.set(
            this.heading.x,
            this.heading.y * Math.cos(angle) - this.heading.z * Math.sin(angle),
            this.heading.y * Math.sin(angle) + this.heading.z * Math.cos(angle)
        );
    }

    /**
     * Rotation around the Y axis
     * @param {Float} angle in radians
     */
    rotY(angle) {
        if (this.debug) {
            logging.log('[Turtle] rotY(' + angle + ')');
        }
        angle = this.randomize(angle);
        this.heading.set(
            this.heading.x * Math.cos(angle) + this.heading.z * Math.sin(angle),
            this.heading.y,
            - this.heading.x * Math.sin(angle) + this.heading.z * Math.cos(angle));
    }

    /**
     * Rotation around the Z axis
     * @param {Float} angle in radians
     */
    rotZ(angle) {
        if (this.debug) {
            logging.log('[Turtle] rotZ(' + angle + ')');
        }
        angle = this.randomize(angle);
        this.heading.set(
            this.heading.x * Math.cos(angle) - this.heading.y * Math.sin(angle),
            this.heading.x * Math.sin(angle) + this.heading.y * Math.cos(angle),
            this.heading.z
        );
    }

    /**
     * Set the color
     * @param {Object} color - {r, g, b} color object
     */
    setColor(color) {
        if (this.debug) {
            logging.log('[Turtle] setColor(' + color + ')');
        }
        this.color.setRGB(
            color.r,
            color.g,
            color.b
        );
    }

    /**
     * Get to the next color, by order of hue
     * @param {Float} step - step by which to increase the hue
     */
    nextColor(step) {
        if (this.debug) {
            logging.log('[Turtle] nextColor(' + step + ')');
        }
        const hsl = this.color.clone().getHSL({});
        let newH = hsl.h + step;
        if (newH >= 1.0) {
            newH -= 1.0;
        }
        this.color.setHSL(
            newH,
            hsl.s,
            hsl.l
        );
    }

    /**
     * End a command sequence
     */
    end() {
    }

    /**
     * Move backward (and draw a point)
     * @param {Float} step distance
     */
    backward(step) {
        return this.forward(-step)
    }

    /**
     * Inverted rotation around the X axis
     * @param {Float} angle in radians
     */
    iRotX(angle) {
        return this.rotX(-angle);
    }

    /**
     * Inverted rotation around the Y axis
     * @param {Float} angle in radians
     */
    iRotY(angle) {
        return this.rotY(-angle);
    }

    /**
     * Inverted rotation around the Z axis
     * @param {Float} angle in radians
     */
    iRotZ(angle) {
        return this.rotZ(-angle);
    }



    _miniSphere() {
        return new THREE.Mesh(
            new THREE.SphereBufferGeometry(0.01, 8, 4),
            new THREE.MeshBasicMaterial({ color: this.color.clone() })
        );
    }
}

class VertexBuffer {
    constructor() {
        this.positions = [];
        this.colors = [];
    }

    addVertex(position, color) {
        for (const component of position) {
            this.positions.push(component);
        }
        for (const component of color) {
            this.colors.push(component);
        }
    }
}

/**
 * Represents a state of the turtle, for push and pop operations
 */
class TurtleState {
    /**
     * Initialize a state of the turtle
     * @param {THREE.Vector3} pos - position of the turtle, THREE.Vector3 object
     * @param {THREE.Vector3} heading - heading of the turtle, THREE.Vector3 object
     * @param {THREE.Color} color - color of the turtle, as a THREE.Color object
     */
    constructor(pos, heading, color) {
        this.pos = pos.clone();
        this.heading = heading.clone();
        this.color = color.clone();
    }
}
